src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataReaderExtensions.cs

```diff
@@ -140,5 +180,37 @@ public static ODataItemBase ReadResourceOrResourceSet(this ODataReader reader)
             Contract.Assert(topLevelItem != null, "A top level resource or resource set should have been read by now.");
             return topLevelItem;
         }
+
+        private static void AddNestedResourceInfo(ref ODataItemBase topLevelItem, Stack<ODataItemBase> itemsStack, ODataResourceSetWrapperBase resourceSetWrapper)
+        {
+            if (itemsStack.Count > 0)
+            {
+                ODataNestedResourceInfoWrapper parentNestedResourceInfo = (ODataNestedResourceInfoWrapper)itemsStack.Peek();
+                Contract.Assert(parentNestedResourceInfo != null, "this has to be an inner resource set. inner resource sets always have a nested resource info.");
+                Contract.Assert(parentNestedResourceInfo.NestedResourceInfo.IsCollection == true, "Only collection nested properties can contain resource set as their child.");
+                parentNestedResourceInfo.NestedItems.Add(resourceSetWrapper);
+            }
+            else
+            {
+                topLevelItem = resourceSetWrapper;
+            }
+        }
+
+        private static void AddResourceToParent(Stack<ODataItemBase> itemsStack, ODataResourceWrapper resourceWrapper)
+        {
+            ODataItemBase parentItem = itemsStack.Peek();
+            ODataResourceSetWrapper parentResourceSet = parentItem as ODataResourceSetWrapper;
+            if (parentResourceSet != null)
+            {
+                parentResourceSet.Resources.Add(resourceWrapper);
+            }
+            else
+            {
+                ODataNestedResourceInfoWrapper parentNestedResource = (ODataNestedResourceInfoWrapper)parentItem;
+                Contract.Assert(parentNestedResource.NestedResourceInfo.IsCollection == false, "Only singleton nested properties can contain resource as their child.");
+                Contract.Assert(parentNestedResource.NestedItems.Count == 0, "Each nested property can contain only one resource as its direct child.");
+                parentNestedResource.NestedItems.Add(resourceWrapper);
```

Comment:
I realize this was the previous logic, but it doesn't make sense to me.  Is this saying that each nested (delta)resourcesetWrapper can only have a single item? That may have been true in the past (since we didn't allow deep inserts/updated), but (unless I misunderstand what this is asserting) I don't think it's true now.

Does this assert get raised when doing a deep insert or update with a collection?

Reply:
Comment seems to have lost context -- see AddResourceToParent, line 232. 

---
In reply to: [518439388](https://github.com/OData/WebApi/pull/2275#discussion_r518439388) [](ancestors = 518439388)